plugin/pathvirt/pathvirt.cpp

```diff
@@ -0,0 +1,240 @@
+    }
+
+    /* not found */
+    return -1;
+}
+
+/*
+ * dynamic_path_swap - given old path, return new path or NULL
+ *
+ * Returns NULL if no swap is to be done and the original path value should
+ * be used. Returns a malloc'd pointer to the new string if a swap should
+ * happen.
+ *
+ * If didn't return NULL, the returned pointer must be freed.
+ */
+bool
+dynamic_path_swap(const char *path, dmtcp::string &newPath)
+{
+    char *oldPathPtr = NULL;
+    /* quickly return NULL if no swap */
+    if (!should_swap) {
+        return false;
+    }
+
+    /* yes, should swap */
+
+    /* check if path is in list of registered paths to swap out */
+    int index = clfind(old_path_prefix_list, path, &oldPathPtr);
+    if (index == -1)
+        return false;
+
+    /* found it in old list, now get a pointer to the new prefix to swap in*/
+    char *newPathPtr = clget(new_path_prefix_list, index);
+    if (newPathPtr == NULL)
+        return false;
+
+    size_t new_element_sz = clgetsize_ptr(new_path_prefix_list, newPathPtr);
+    size_t old_element_sz = clgetsize_ptr(old_path_prefix_list, oldPathPtr);
+
+    /* temporarily null terminate new element */
+    newPathPtr[new_element_sz] = '\0';
+
+    /* finally, create full path with the new prefix swapped in */
+
+    /* plus 1 is for safety slash we include between the new prefix and the
+       unchanged rest of the path. this is in case their environment
+       variable doesn't end with a slash. in the "worst" case,
+       there will be two extra slashes if the new prefix ends with a slash
+       and the old one doesn't. plus 1 for NULL */
+    size_t newpathsize = (strlen(path) - old_element_sz) + new_element_sz + 1 + 1;
+    newPath = newPathPtr;
+    newPath += "/";
+    newPath += (path + old_element_sz);
+
+    /* repair the colon list */
+    newPathPtr[new_element_sz] = ':';
+
+    return true;
+}
+
+/*
+ * DMTCP Setup
+ */
+
+void
+dmtcp_event_hook(DmtcpEvent_t event, DmtcpEventData_t *data)
+{
+    /* NOTE:  See warning in plugin/README about calls to printf here. */
+    switch (event) {
+    case DMTCP_EVENT_INIT:
+    {
+        /* On init, check if they've specified paths to virtualize via
+           DMTCP_PATH_PREFIX env */
+        char *old_env = getenv(ENV_DPP);
+        if (old_env) {
+            /* if so, save it to buffer */
+            snprintf(old_path_prefix_list, sizeof(old_path_prefix_list), "%s",
+                     old_env);
+        }
+        break;
+    }
+    case DMTCP_EVENT_RESTART:
+    {
+        /* necessary since we don't know how many bytes dmtcp_get_restart_env
+           will write */
+        memset(new_path_prefix_list, 0, sizeof(new_path_prefix_list));
+
+        /* Try to get the value of ENV_DPP from new environment variables,
+         * passed in on restart */
+        int ret = dmtcp_get_restart_env(ENV_DPP, new_path_prefix_list,
+                                        sizeof(new_path_prefix_list) - 1);
+
+        JASSERT(ret == 0);
```

Comment:
@mossberg, We should document what the behavior is clearly in a README or comments.

I would propose the following:  `DMTCP_PATH_PREFIX` is never required on restart.  If  `DMTCP_PATH_PREFIX` was not provided on the current restart, then there will _not_ be any path virtualization.

It seems to me that this is a feature rather than a bug.  :-)  A user can decide much later at restart time if they need path virtualization.  For example, maybe they need it if restarting on a new computer, and don't need it if restarting onthe original computer.